src/Elliptic/BoundaryConditions/AnalyticSolution.hpp

```diff
@@ -0,0 +1,234 @@
+ * as Dirichlet or Neumann boundary conditions for each field separately.
+ * Dirichlet boundary conditions are imposed on the fields and Neumann boundary
+ * conditions are imposed on the fluxes.
+ *
+ * See `elliptic::Actions::InitializeAnalyticSolutions` for an action that can
+ * add the analytic solutions to the DataBox.
+ */
+template <typename System, size_t Dim, typename... FieldTags,
+          typename... FluxTags, typename Registrars>
+class AnalyticSolution<System, Dim, tmpl::list<FieldTags...>,
+                       tmpl::list<FluxTags...>, Registrars>
+    : public BoundaryCondition<Dim, Registrars> {
+ private:
+  using Base = BoundaryCondition<Dim, Registrars>;
+
+ public:
+  using options =
+      tmpl::list<elliptic::OptionTags::BoundaryConditionType<FieldTags>...>;
+  static constexpr Options::String help =
+      "Boundary conditions from the analytic solution";
+
+  AnalyticSolution() = default;
+  AnalyticSolution(const AnalyticSolution&) noexcept = default;
+  AnalyticSolution& operator=(const AnalyticSolution&) noexcept = default;
+  AnalyticSolution(AnalyticSolution&&) noexcept = default;
+  AnalyticSolution& operator=(AnalyticSolution&&) noexcept = default;
+  ~AnalyticSolution() noexcept = default;
+
+  /// \cond
+  explicit AnalyticSolution(CkMigrateMessage* m) noexcept : Base(m) {}
+  using PUP::able::register_constructor;
+  WRAPPED_PUPable_decl_template(AnalyticSolution);
+  /// \endcond
+
+  /// Select which `elliptic::BoundaryConditionType` to apply for each field
+  explicit AnalyticSolution(
+      const typename elliptic::OptionTags::BoundaryConditionType<
+          FieldTags>::type... boundary_condition_types) noexcept
+      : boundary_condition_types_{boundary_condition_types...} {}
+
+  std::unique_ptr<domain::BoundaryConditions::BoundaryCondition> get_clone()
+      const noexcept override {
+    return std::make_unique<AnalyticSolution>(*this);
+  }
+
+  const auto& boundary_condition_types() const noexcept {
+    return boundary_condition_types_;
+  }
+
+  using argument_tags =
+      tmpl::list<::Tags::AnalyticSolutionsBase, domain::Tags::Mesh<Dim>,
+                 domain::Tags::Direction<Dim>,
+                 ::Tags::Normalized<domain::Tags::UnnormalizedFaceNormal<
+                     Dim, Frame::Inertial>>>;
+  using volume_tags =
+      tmpl::list<::Tags::AnalyticSolutionsBase, domain::Tags::Mesh<Dim>>;
+
+  template <typename OptionalAnalyticSolutions>
+  void apply(const OptionalAnalyticSolutions& optional_analytic_solutions,
+             const Mesh<Dim>& volume_mesh, const Direction<Dim>& direction,
+             const tnsr::i<DataVector, Dim>& face_normal,
+             const gsl::not_null<typename FieldTags::type*>... fields,
+             const gsl::not_null<typename FieldTags::type*>... n_dot_fluxes)
+      const noexcept {
+    const auto& analytic_solutions =
+        [&optional_analytic_solutions]() noexcept -> decltype(auto) {
```

Comment:
why do you need to specify `decltype(auto)` as the return type?

Reply:
Without the `decltype(auto)` the lambda assumes the return type is `auto` and returns by value. With the `decltype(auto)` the lambda returns by const-ref. This is assuming that I got this C++ rule right.

Reply:
Okay, would `const auto&`as a return type work? That might be clearer

Reply:
Yep that works. I think `decltype(auto)` is the right choice for generic code where the return value should be forwarded exactly.